### PR TITLE
[Performance] Use segment operators for graph readout.

### PR DESCRIPTION
**Pull request by yzh119**
## Description
Previously we created a pseudo graph and applied SpMM to get graph readout results, and introduced graph creation overhead. Thus PR implements segment operators to lower this (and corresponding coo/csr conversion) overhead.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/backend/backend.py [hunk 1]
```diff
@@ -1489,6 +1506,31 @@ def edge_softmax(gidx, logits, eids, norm_by):
...
+    Zero-length segments are allowed.
+
+    Parameters
+    ----------
+    op : str
+        Aggregation method. Can be 'sum', 'max', 'min'.
+    seglen : Tensor
+        Segment lengths.
+    value : Tensor
+        Value to aggregate.
```
Comment (jermainewang) [inline python/dgl/backend/backend.py:1526]: The docstring is inconsistent with the function signature.

File: python/dgl/backend/pytorch/sparse.py [hunk 2]
```diff
@@ -231,6 +231,32 @@ def backward(ctx, grad_out):
         return None, grad_score, None, None
 
 
+class SegmentReduce(th.autograd.Function):
+    @staticmethod
+    def forward(ctx, op, x, offsets):
+        y, arg = _segment_reduce(op, x, offsets)
+        print(arg)
```
Comment (jermainewang) [inline python/dgl/backend/pytorch/sparse.py:238]: print?

File: python/dgl/backend/backend.py (hunk 1 above)
Comment (jermainewang) [reply python/dgl/backend/backend.py:1526]: The docstring misses many important descriptions. For example, is the offsets zero-based or one-based?

File: python/dgl/sparse.py [hunk 3]
```diff
@@ -248,4 +248,81 @@ def _gsddmm(gidx, op, lhs, rhs, lhs_target='u', rhs_target='v'):
...
+    Zero-length segments are allowed.
+
+    Parameters
+    ----------
+    op : str
+        Aggregation method. Can be 'sum', 'max', 'min'.
+    seglen : Tensor
+        Segment lengths.
+    value : Tensor
+        Value to aggregate.
```
Comment (jermainewang) [inline python/dgl/sparse.py:266]: The docstring is totally wrong

File: src/array/cpu/segment_reduce.h [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  for (int i = 1; i < out->ndim; ++i)
+    dim *= out->shape[i];
+  const DType* feat_data = feat.Ptr<DType>();
+  const IdType* offsets_data = offsets.Ptr<IdType>();
+  DType *out_data = out.Ptr<DType>();
+#pragma omp parallel for
+  for (int i = 0; i < n; ++i) {
+    for (IdType j = offsets_data[i]; j < offsets_data[i + 1]; ++j) {
+      for (int k = 0; k < dim; ++k) {
+        out_data[i * dim + k] += feat_data[j * dim + k];
```
Comment (jermainewang) [inline src/array/cpu/segment_reduce.h:28]: Use a local variable for accum

File: src/array/cpu/segment_reduce.h [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  std::fill(out_data, out_data + out.NumElements(), Cmp::zero);
+  std::fill(arg_data, arg_data + arg.NumElements(), -1);
+#pragma omp parallel for
+  for (int i = 0; i < n; ++i) {
+    for (IdType j = offsets_data[i]; j < offsets_data[i + 1]; ++j) {
+      for (int k = 0; k < dim; ++k) {
+        const DType val = feat_data[j * dim + k];
+        if (Cmp::Call(out_data[i * dim + k], val)) {
+          out_data[i * dim + k] = val;
+          arg_data[i * dim + k] = j;
```
Comment (jermainewang) [inline src/array/cpu/segment_reduce.h:54]: the same as above

File: src/array/cuda/functor.cuh [hunk 6]
```diff
@@ -133,6 +135,15 @@ struct Sum {
...
   }
+  static __device__ __forceinline__ void Call(
+      DType *out_buf, Idx *arg_buf,
+      DType val, Idx id) {
+    if (!atomic) {
+      *out_buf += val;
+    } else {
+      cuda::AtomicAdd(out_buf, val);
+    }
+  }
```
Comment (jermainewang) [inline src/array/cuda/functor.cuh:146]: We should refactor this part.

File: src/array/cuda/segment_reduce.cuh [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+namespace aten {
+namespace cuda {
+
+/*!
+ * \brief CUDA kernel of segment reduce.
+ */
+template <typename IdType, typename DType,
+          typename ReduceOp>
+__global__ void SegmentReduceKernel(
```
Comment (jermainewang) [inline src/array/cuda/segment_reduce.cuh:25]: SegmentReduceKernel -> SegmentReduceCmpKernel

Also, poor docstring

**Comment by jermainewang** [comment]
Also missing C++ tests

**Comment by yzh119** [comment]
Will fix the documentation issue soon, for other concerns:
1. `print` has been removed in #2362.
2. We found that on CPU (#2201 ), the loop order (graph dim, feature dim) is far better than (feature dim, graph dim) , in the later case we don't need an accumulative variable.
